test/metabase/query_processor/middleware/cache_test.clj

```diff
@@ -323,7 +327,32 @@
                        (-> cached-result
                            (dissoc :cached :updated_at)
                            (m/dissoc-in [:data :results_metadata :checksum])))
-                    "Cached result should be in the same format as the uncached result, except for added keys")))))))))
+                    "Cached result should be in the same format as the uncached result, except for added keys"))))))))
+  (testing "Cached results don't impact average execution time"
+    (let [query                         (assoc (mt/mbql-query venues {:order-by [[:asc $id]] :limit 42})
+                                               :cache-ttl 5000)
+          q-hash                        (qputil/query-hash query)
+          call-count                    (atom 0)
+          called-promise                (promise)
+          save-query-execution-original (var-get #'process-userland-query/save-query-execution!*)]
+      (with-redefs [process-userland-query/save-query-execution!* (fn [& args]
+                                                                    (swap! call-count inc)
+                                                                    (apply save-query-execution-original args)
+                                                                    (deliver called-promise true))
+                    cache/min-duration-ms                         (constantly 0)]
+        (with-mock-cache [save-chan]
+          (db/delete! Query :query_hash q-hash)
+          (is (not (:cached (qp/process-userland-query query (context.default/default-context)))))
+          (a/alts!! [save-chan (a/timeout 200)]) ;; wait-for-result closes the channel
+          (u/deref-with-timeout called-promise 500)
```

Comment:
this one flaked for me over three tries testing.

Reply:
can you try finding better values for the `200` and `500` numbers? I can see if there is some decent way to make it more reliable. But fundamentally I need things to exist in a cache, and have timings in the db, both of which happen asynchronous to the query thread.

Reply:
I tried 10 times at `200` and `800` in the same settings and it hasn't flaked yet